mlx/backend/metal/kernels/utils.h

```diff
@@ -288,12 +324,110 @@ METAL_FUNC uint3 elem_to_loc_3_nd(
   return loc;
 }
 
+///////////////////////////////////////////////////////////////////////////////
+// Elem to loc in a loop utils
+///////////////////////////////////////////////////////////////////////////////
+
+template <int dim, typename offset_t = size_t>
+struct looped_elem_to_loc {
+  looped_elem_to_loc<dim - 1, offset_t> inner_looper;
+  offset_t offset{0};
+  int index{0};
+
+  void next(const constant int* shape, const constant size_t* strides) {
+    index++;
+    offset += strides[dim - 1];
+
+    if (index >= shape[dim - 1]) {
+      index = 0;
+      inner_looper.next(shape, strides);
+      offset = inner_looper.offset;
+    }
+  }
+
+  void next(int n, const constant int* shape, const constant size_t* strides) {
+    index += n;
+    offset += n * strides[dim - 1];
+
+    if (index >= shape[dim - 1]) {
+      int extra = index - shape[dim - 1];
+      index = 0;
+      inner_looper.next(shape, strides);
+      offset = inner_looper.offset;
+      if (extra > 0) {
+        next(extra, shape, strides);
+      }
+    }
+  }
+
+  offset_t location(
+      offset_t idx,
+      const constant int* shape,
+      const constant size_t* strides,
+      int ndim) {
+    (void)idx; // appease the compiler
+    (void)shape;
+    (void)strides;
+    (void)ndim;
+    return offset;
+  }
+};
+
+template <typename offset_t>
+struct looped_elem_to_loc<1, offset_t> {
+  offset_t offset{0};
+
+  void next(const constant int* shape, const constant size_t* strides) {
+    (void)shape; // appease the compiler
+    offset += strides[0];
+  }
+
+  void next(int n, const constant int* shape, const constant size_t* strides) {
+    (void)shape; // appease the compiler
+    offset += n * strides[0];
+  }
+
+  offset_t location(
+      offset_t idx,
+      const constant int* shape,
+      const constant size_t* strides,
+      int ndim) {
+    (void)idx; // appease the compiler
+    (void)shape;
+    (void)strides;
+    (void)ndim;
+    return offset;
+  }
```

Comment:
Nit: you can do that more simply by leaving out the names, e.g.

```suggestion
  offset_t location(
      offset_t idx,
      const constant int*,
      const constant size_t*,
      int) {
    return offset;
  }
```

Reply:
Awesome! Thanks, I hated having to write those `(void)` things.